test/UnitTests/Microsoft.Testing.Platform.UnitTests/Configuration/ConfigurationManagerTests.cs

```diff
@@ -0,0 +1,179 @@
+                fileSystem.Object, null));
+
+        IConfiguration configuration = await configurationManager.BuildAsync(
+            serviceProviderMock.Object,
+            loggerProviderMock.Object);
+        Assert.AreEqual(result, configuration[key], $"Expected '{result}' found '{configuration[key]}'");
+
+        loggerMock.Verify(x => x.LogAsync(LogLevel.Trace, It.IsAny<string>(), null, LoggingExtensions.Formatter), Times.Once);
+    }
+
+    public async ValueTask BuildAsync_EmptyConfigurationSources_ThrowsException()
+    {
+        Mock<ServiceProvider> mockServiceProvider = new();
+
+        ConfigurationManager configurationManager = new();
+        await Assert.ThrowsAsync<InvalidOperationException>(
+            () => configurationManager.BuildAsync(
+                mockServiceProvider.Object,
+                null));
+    }
+
+    public async ValueTask BuildAsync_ConfigurationSourcesNotEnabledAsync_ThrowsException()
+    {
+        Mock<IConfigurationSource> mockConfigurationSource = new();
+        mockConfigurationSource.Setup(x => x.IsEnabledAsync()).ReturnsAsync(false);
+
+        Mock<ServiceProvider> mockServiceProvider = new();
+
+        ConfigurationManager configurationManager = new();
+        configurationManager.AddConfigurationSource(() => mockConfigurationSource.Object);
+
+        await Assert.ThrowsAsync<InvalidOperationException>(
+            () => configurationManager.BuildAsync(
+                mockServiceProvider.Object,
+                null));
+
+        mockConfigurationSource.Verify(x => x.IsEnabledAsync(), Times.Once);
+    }
+
+    public async ValueTask BuildAsync_ConfigurationSourceIsAsyncInitializableExtension_InitializeAsyncIsCalled()
+    {
+        Mock<IConfigurationProvider> mockConfigurationProvider = new();
+        mockConfigurationProvider.Setup(x => x.LoadAsync()).Callback(() => { });
+
+        Mock<ServiceProvider> mockServiceProvider = new();
+
+        Mock<FakeConfigurationSource> fakeConfigurationSource = new();
+        fakeConfigurationSource.Setup(x => x.IsEnabledAsync()).ReturnsAsync(true);
+        fakeConfigurationSource.Setup(x => x.InitializeAsync()).Callback(() => { });
+        fakeConfigurationSource.Setup(x => x.Build()).Returns(mockConfigurationProvider.Object);
+
+        ConfigurationManager configurationManager = new();
+        configurationManager.AddConfigurationSource(() => fakeConfigurationSource.Object);
+
+        await Assert.ThrowsAsync<InvalidOperationException>(
+            () => configurationManager.BuildAsync(
+                mockServiceProvider.Object,
+                null));
+
+        fakeConfigurationSource.Verify(x => x.IsEnabledAsync(), Times.Once);
+        fakeConfigurationSource.Verify(x => x.InitializeAsync(), Times.Once);
+    }
+}
+
+#pragma warning disable CA1852
+internal class FakeConfigurationSource : IConfigurationSource, IAsyncInitializableExtension
```

Comment:
Why not seal the class instead of disabling the rule?

Reply:
We cannot seal it because we have virtual members.